xayn-ai/src/reranker/sync.rs

```diff
@@ -0,0 +1,157 @@
+        Ok(data)
+    }
+
+    /// Serializes a `SyncData` to a byte representation.
+    pub(crate) fn serialize(&self) -> Result<Vec<u8>, Error> {
+        let size = bincode::serialized_size(self)? + 1;
+        let mut serialized = Vec::with_capacity(size as usize);
+        // version encoded in first byte
+        serialized.push(CURRENT_SCHEMA_VERSION);
+        bincode::serialize_into(&mut serialized, self)?;
+
+        Ok(serialized)
+    }
+
+    /// Synchronizes with another `SyncData`.
+    pub(crate) fn synchronize(&mut self, other: SyncData) {
+        let Self { user_interests } = other;
+        self.user_interests.append(user_interests);
+
+        reduce_cois(&mut self.user_interests.positive);
+        reduce_cois(&mut self.user_interests.negative);
+
+        self.user_interests.reassign_ids();
+    }
+}
+
+/// A pair of CoIs.
+#[derive(Clone)]
+struct CoiPair<C>(C, C);
+
+impl<C> CoiPair<C>
+where
+    C: CoiPoint,
+{
+    /// Merges the CoI pair, assigning it the smaller of the two ids.
+    fn merge(&self) -> C {
+        let CoiId(id1) = self.0.id();
+        let CoiId(id2) = self.1.id();
+        self.0.merge(&self.1, id1.min(id2))
+    }
+
+    /// True iff either CoI has the given id.
+    fn contains(&self, id: CoiId) -> bool {
+        self.0.id() == id || self.1.id() == id
+    }
+
+    /// True iff either CoI is one of the given pair.
+    fn contains_any(&self, other: &Self) -> bool {
+        self.contains(other.0.id()) || self.contains(other.1.id())
+    }
+}
+
+/// A `Coiple` is a pair of CoIs and the distance between them.
+#[derive(Clone)]
+struct Coiple<C> {
+    cois: CoiPair<C>,
+    dist: f32,
+}
+
+impl<C> Coiple<C> {
+    fn new(coi1: C, coi2: C, dist: f32) -> Self {
+        let cois = CoiPair(coi1, coi2);
+        Self { cois, dist }
+    }
+}
+
+/// Computes the l2 distance between two CoI points.
+fn dist<C>(coi1: &C, coi2: &C) -> f32
+where
+    C: CoiPoint,
+{
+    l2_distance(coi1.point(), coi2.point())
+}
+
+/// Reduces the given collection of CoIs by successively merging the pair in closest
+/// proximity to each other.
+///
+/// <https://xainag.atlassian.net/wiki/spaces/XAY/pages/2029944833/CoI+synchronisation>
+/// outlines the core of the algorithm.
+fn reduce_cois<C>(cois: &mut Vec<C>)
+where
+    C: CoiPoint + Clone,
+{
+    // initialize collection of close coiples
+    let cois_iter = cois.iter();
+    let mut coiples = cois_iter
+        .clone()
+        .cartesian_product(cois_iter)
+        .filter(|(coi1, coi2)| coi1.id() < coi2.id())
+        .filter_map(|(coi1, coi2)| {
+            let dist = dist(coi1, coi2);
+            (dist < SOCIAL_DIST).then(|| Coiple::new(coi1.clone(), coi2.clone(), dist))
+        })
+        .collect_vec();
+
+    while !coiples.is_empty() {
+        // find the minimum i.e. closest coiple
+        let min_coiple = coiples
+            .iter()
+            .cloned()
+            .min_by(|cpl1, cpl2| nan_safe_f32_cmp_high(&cpl1.dist, &cpl2.dist))
```

Comment:
In case we have multiple distances with the same value (in the case most of the time the distance will just be `0`) we need to give them an order based on same other field. On confluence we went for `alpha` and `beta` but negative cois do not have them but we can use the id.

```suggestion
            .min_by(|cpl1, cpl2| nan_safe_f32_cmp_high(&cpl1.dist, &cpl2.dist)
                                 .then_with(|| cpl1.id().cmp(cpl2.id()))
```

Reply:
I just realized that this is wrong because `cpl*` contains two cois, so this will need to be more elaborate.

Reply:
thanks, I had overlooked this case. I've now added a comparison over coiples that does a lexicographic comparison over the component coi ids.